iceoryx_utils/include/iceoryx_utils/cxx/sized_uninitialized_array.hpp

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2021 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ICEORYX_UTILS_CXX_SIZED_UNINITIALIZED_ARRAY_HPP
+#define ICEORYX_UTILS_CXX_SIZED_UNINITIALIZED_ARRAY_HPP
+
+#include "iceoryx_utils/cxx/sized_container.hpp"
+#include <cstdint>
+
+namespace iox
+{
+namespace cxx
+{
+template <typename T, uint64_t Capacity>
+class SizedUninitializedArray : public SizedContainer<T, Capacity>
+{
+  public:
+    using value_type = T;
+    using size_type = decltype(Capacity);
+
+    SizedUninitializedArray() = default;
+
+    /// @brief return the pointer to the underlying array
+    /// @return pointer to underlying array
+    ///          If the vector is empty it returns nullptr
+    T* data() noexcept;
+
+    /// @brief return the const pointer to the underlying array
+    /// @return const pointer to underlying array
+    const T* data() const noexcept;
+
+
+    /// @brief returns the capacity, which was given via the template argument,
+    ///         i.e., the maximum number of elements it can hold
+    uint64_t capacity() const noexcept;
+
+    /// @brief returns the capacity, which was given via the template argument,
+    ///         i.e., the maximum number of elements it can hold (same as capacity())
+    uint64_t max_size() const noexcept;
+
+
+  protected:
+    using element_t = uint8_t[Capacity == 0 ? 1 : sizeof(T)];
+
+  private:
+    alignas(Capacity == 0 ? 1 : alignof(T)) element_t m_data[Capacity == 0 ? 1 : Capacity];
```

Comment:
We have to carefully think about possible issues arising from this. I have a slight leaning towards having this aligned to `T`.
The `std::array` also keeps the alignment of `T` if the capacity is 0 and I would rather follow the STL and trust their judgement.

Reply:
@elBoberido I just tried `std::array<int, 0> x` with libstdc++ 7.3.0 and g++ 7.5, which gives me `sizeof(x)==1`, `alignof(x)==1`, but `alignof(int)==4`. So I could not reproduce this. Which versions do you use? 
The explicit alignas in the uninitialized_array is just needed because the uint8_t array is later cast to T. And accessing a wrongly aligned T has to be prevented. But an empty array is never cast to T anyways, we even do not hold enough memory for a single T (assuming sizeof(T)>1). Always aligning to T would potentially waste memory if Capacity is 0.

Reply:
.@sschnitzer you are right, I must have hallucinated. Like I said, it's just a slight leaning towards aligning it to `T`, since it also affects the code surrounding it, but it's better to stick with the STL and set the alignment to 1.

Just one thing, since `element_t` is protected, it might be better to do the alignment on `element_t` instead of `m_data`.

Reply:
Doing the alignment on element_t is a good idea in case it is used in some other class. Will change that.